brazilfiscalreport/dacte/dacte.py

```diff
@@ -1100,94 +1100,111 @@ def _draw_service_recipient(self, config):
+        remaining_width = page_width - (x_margin + 2) - cubagem_width
+        other_col_width = remaining_width / 4
 
-        self.set_xy(x_line_2, section_start_y)
-        self.set_font(self.default_font, "", 7)
-        self.multi_cell(w=0, h=4, text="VALOR TOTAL DA MERCADORIA", align="L")
-        self.set_xy(x_line_2, section_start_y)
-        self.set_font(self.default_font, "B", 7)
-        self.multi_cell(w=0, h=14, text=f"R$ {self.inf_carga_valor}", align="L")
-
-        section_start_y += 10
+        # Calcula as posições X para cada coluna
+        x_line_1 = x_margin + other_col_width
+        x_line_2 = x_line_1 + other_col_width
+        x_line_3 = x_line_2 + other_col_width
+        x_line_4 = x_line_3 + cubagem_width  # Posição após o campo de cubagem
 
-        self.rect(
-            x=x_margin, y=section_start_y, w=page_width - 0.1 * x_margin, h=10, style=""
-        )
-        col_width = (page_width - (x_margin + 2)) / 4
-        x_line_1 = x_margin - 55 + col_width
-        x_line_2 = x_margin - 30 + 2 * col_width
-        x_line_3 = x_margin - 40 + 3 * col_width
-        x_line_4 = x_margin - 40 + 4 * col_width
+        # Desenha as linhas verticais
         self.line(
-            x1=x_line_1 + 34,
-            x2=x_line_1 + 34,
+            x1=x_line_1,
+            x2=x_line_1,
             y1=section_start_y,
             y2=section_start_y + 10,
         )
         self.line(x1=x_line_2, x2=x_line_2, y1=section_start_y, y2=section_start_y + 10)
         self.line(x1=x_line_3, x2=x_line_3, y1=section_start_y, y2=section_start_y + 10)
         self.line(x1=x_line_4, x2=x_line_4, y1=section_start_y, y2=section_start_y + 10)
 
-        x_positions = [0, x_line_1 + 34, x_line_2, x_line_3, x_line_4]
-        titles = [
-            "PESO (Kg)",
-            "TP MED /UN. MED",
-            "TP MED /UN. MED",
-            "CUBAGEM (M³)",
-            "QUANTIDADE DE VOLUMES",
+        # Define as posições X e larguras para todos os campos
+        x_positions = [x_margin, x_line_1, x_line_2, x_line_3, x_line_4]
+        col_widths = [
+            other_col_width,
+            other_col_width,
+            other_col_width,
+            cubagem_width,
+            other_col_width,
         ]
 
+        # Campos fixos para cubagem e volumes
+        fixed_titles = ["CUBAGEM (M³)", "QUANTIDADE DE VOLUMES"]
+
+        # Inicializa lista de títulos com espaços vazios para os campos dinâmicos
+        titles = ["", "", ""] + fixed_titles
+
+        # Preenche os títulos dinâmicos baseado nos dados do XML
+        dynamic_fields = []
+        for item in self.inf_carga_list:
+            c_unid, tp_media, q_carga = item
+            if (
+                q_carga
+                and float(q_carga) > 0
+                and len(dynamic_fields) < 3  # Limita a 3 campos dinâmico
+                and c_unid in TP_CODIGO_MEDIDA
+            ):
+                field_title = f"{tp_media} ({TP_CODIGO_MEDIDA[c_unid]})"
```

Comment:
```suggestion
              unidade = TP_CODIGO_MEDIDA[c_unid]
              field_title = tp_media if tp_media == unidade else f"{tp_media} ({unidade})"
```

Aqui acho que poderia algo assim, pra evitar imprimir o titulo "M3 (M3)" e imprimir apenas M3 nos casos que os dois são iguais, o que acha?

Reply:
@leogregianin 
Nesse caso de exemplo, alias a cubagem aparece duas vezes, será que é isso mesmo?
![Imagem do WhatsApp de 2025-04-24 à(s) 16 43 10_3328216b](https://github.com/user-attachments/assets/55047fe2-7665-41f0-a6ef-d1929aa44c35)

Reply:
Temos alguns problemas ai sim, e precisamos definir algumas coisas.

Veja que as DACTes atuais do projeto já existem essas repetições de Cubagem M3 e Unidade.

A tabela de Unidade de medida:
```
00 - M3;
01 - KG;
02 - TON;
03 - UNIDADE;
04 - LITROS;
05 - MMBTU.
```

Podemos definir que `00-M3` sempre será impresso no 4º campo "CUBAGEM M3", que `03-UNIDADE` sempre será impresso no 5º campo "QUANTIDADE DE VOLUMES". O restante dos campos são impressos nos 3 primeiros campos.

Reply:
@leogregianin 

Obrigado por me ajudar a esclarecer e entender melhor essa questão.

Será que não é valido deixar todos os quadros de quantidade dinamico?
Olhando o manual vejo que podemos ter de 1 a "n" quadros de quantidades.
se acontecer de ter mais 5 quadros penso que vamos precisar continuar na linha de baixo.
na DANFE eu já fiz uma logica parecida para imprimir as duplicatas.

apenas para referencia/consulta to anexando essa parte da documentação:
![image](https://github.com/user-attachments/assets/67c8c652-fa79-4f83-834d-1f7089561080)

não precisamos suportar todas as possibilidade de inicio, procurando apenas entender melhor mesmo.

Reply:
Também penso o ideal ser dinâmico, tudo que tiver no XML imprimir alí e continuando na linha abaixo se for o caso de mais de 5 unidades informadas.

Mas também tem que considerar o DACTE publicado no [portal do CTe](https://dfe-portal.svrs.rs.gov.br/Cte/Documentos#) com o nome `MOC CTe 4.00 Anexo II - DACTE` - `Manual de Orientações ao Contribuinte do CTe versão 4.00 - Anexo II DACTE`

Nesse manual consta informações padrões que devem ser impresso no DACTE.

Reply:
@leogregianin 

Continuando a analise, consultei o manual da DACTE confome vc mencionou, na parte de correlação de campos é mostrado assim:
![image](https://github.com/user-attachments/assets/e1584727-d023-45ac-af44-8bd335f93f12)
O que dá pra entender é que se exister essas informações, precisamos converter a unidade de medida conforme está definido ali, por exemplo se no xml o peso está em gramas ou toneladas, na pdf precisamos converter e apresentar em Quilos.

Só que o que não fica claro pra mim, é que ali no manual não é feito uma correlação com o tipo de medida ("tpMed")
Como vamos saber qual "tpMed" é pra usar quando existir mais de uma com o mesmo "cUnid?

Vou tentar ver como o pessoal lidou com isso em outros projetos open source, talvez no ACBR tenha algum material.

Reply:
O DACTe do ACBr fica dessa forma:

![acbr](https://github.com/user-attachments/assets/34a6569c-c8f4-40d1-a25f-fe9f1653cd75)

Olhando para o código:
https://svn.code.sf.net/p/acbr/code/trunk2/Fontes/ACBrDFe/ACBrCTe/DACTE/Fast/ACBrCTeDACTEFR.pas

procedure TACBrCTeDACTEFR.CarregaVolumes;

O 4º e 5º campos são fixos para cubagem e volumes.
Mas também são impressos nas três primeiras colunas três itens para cada coluna, no total, 9 itens.

Não há nada sobre conversão de tipos.

Reply:
é o ACBR imprime a cubagem duas vezes, memso estando zerado 😆 

Mas enfim, se não ha regressões eu acho que podemos seguir com o merge, ou você acha que cabe melhoria? 
qualquer coisa podemos deixar uma issue para ser resolvido posteriormente.

Reply:
Da forma que esse PR está, somente imprime 3 itens e ainda com as repetições de cubagem e unidade. Deixamos dessa forma ou fazemos exatamente como o ACBr faz?